E2E: add test case to upload image using Calypso Media modal.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds a test step in the `Blocks: Media (Upload)` spec to test functionality of the Calypso Media modal.

Context: pciE2j-sq-p2

Key changes:
- addition of methods and selectors in `ImageBlock` to support the use of calypso media modal screen.
- addition of method in `MediaPage` to handle scenario of the page being embedded in a modal.
- minor changes to supporting methods as required.

#### Testing instructions
 
- [x] Ensure for mobile and desktop:
  - [x] Blocks: Media (Upload) passes.
  - [x] New test step `upload image file using Calypso media modal` passes.


Related to #

worldomonation committed:
Click on the inserted block once to dismiss the block inserter.

Add test case to cover Calypso Media modal.

Put closure of blockInserter in `GutenbergEditorPage` behind a viewport check.

worldomonation committed:
Update selector used for `GutenbergEditorPage.viewButton`.
Add steps to wait for the full rendering of OpenTable block prior to publishing.
Add some more comments on why the block inserter panel should be closed.

worldomonation commented on packages/calypso-e2e/src/lib/blocks/block-flows/opentable.ts:
```diff
@@ -44,6 +45,15 @@ export class OpenTableFlow implements BlockFlow {
+		// Wait until the block is fully rendered prior to next steps such as publishing.
+		// Alternatively, save as draft works here.
+		// See: https://github.com/Automattic/wp-calypso/issues/50302
+		const openTableFrameHandle = await context.editorIframe.waitForSelector(
+			`${ blockParentSelector } iframe`
+		);
+		const openTableFrame = ( await openTableFrameHandle.contentFrame() ) as Frame;
+		await openTableFrame.waitForSelector( selectors.openTableLogo );
```
As noted in the code comments - an alternative is simply to perform `gutenbergEditorPaage.saveDraft`, which is the standard bypass method for https://github.com/Automattic/wp-calypso/issues/50302.

dpasque commented on packages/calypso-e2e/src/lib/blocks/image-block.ts:
```diff
@@ -63,6 +72,40 @@ export class ImageBlock {
+	 * @param {Sources} source Source for the image.
+	 */
+	async selectImageSource( source: Sources ): Promise< void > {
+		const buttonHandle = await this.block.waitForSelector( selectors.selectImageSourceButton );
+		await buttonHandle.click();
```
Very small nit, but since you're using the `locator` API in the next couple lines, should we use that here too? It also tends to be more stable than using element handles.

worldomonation replied:
Good call, thanks!

worldomonation replied:
Okay, so looks like it is not possible to use the `locator` call from an ElementHandle, which is what `this.block` boils down to.

dpasque replied:
Ahhh right, good point! All good 👍

dpasque commented on packages/calypso-e2e/src/lib/blocks/image-block.ts:
```diff
@@ -63,6 +72,40 @@ export class ImageBlock {
+	 * @param {string} path Path to the image file.
+	 */
+	async uploadFromModal( path: string ): Promise< ElementHandle > {
+		const frame = ( await this.block.ownerFrame() ) as Frame;
+		const page = frame.page();
+
+		// The resulting React modal is essentially the MediaPage, just
+		// embedded within the modal, with additional buttons to
+		// confirm or cancel image insertion.
+		const mediaPage = new MediaPage( page );
+		await mediaPage.upload( path );
+		await mediaPage.clickModalButton( 'Confirm' );
+		await this.waitUntilUploaded();
+		return await this.getImage();
```
This could/should probably be a different PR, but what do you think of moving these media blocks over to the block flow smoke test format soon? 

I was just thinking that it seems like a lot of work here to try to get the page, and moving over the `BlockFlow` interface might make that a little simpler! It also might be good to eventually break the dependence on passing around the element handle for the block.

worldomonation replied:
I intentionally didn't want to migrate everything to the block testing format to keep the scope of this PR tight. With that said, how would one implement all sorts of disparate actions with the block testing format?

In my mind, the block testing flow works because we're only performing one or two actions out of all possible actions on the block. Furthermore, the data that's input and the actions performed on the block themselves are on 'rails' so to speak.

In contrast, in these tests we're looking to test various aspects of the image upload process. 

I suppose the different actions can be handled in the block flow by placing multiple `if/else` branches?

My primary concern is that by using the block testing flow, we lose the flexibility to string together actions against the block to run many different scenarios.

dpasque replied:
That's a good point, and I definitely think any changes would be out of the scope of this PR!

So, my design idea was that I think we can trim down the number of test cases on the Image Block and still get roughly the same coverage. The way I see it, some of the cases could potentially be considered supersets of the others. So, for example, if we just ran the case where we uploaded a reserved file name through the Calypso media modal, I would feel pretty confident in the overall functioning of the block without doing a direct upload on the editor (which is covered in other media based blocks) and that a non-reserved name also works (because we only have special handling for the reserved name).

So, in this case, the "configure" function for the image block just uploads an image with a reserved name through the Calypso media modal. 

What do you think? I'd also be glad to do the follow-up PR for this! 😄

worldomonation replied:
Having thought over this a bit, I agree mostly about there being some duplication in the test.

I'd characterize the following as distinct and thus requiring coverage at this time:
- insert media: calypso modal
- insert media: direct upload

I'll spend some time thinking about best way to do this. It will likely come down to a if/else branching in the `configure` method.

worldomonation replied:
Filed a new task at https://github.com/Automattic/wp-calypso/issues/61140.

dpasque commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -173,4 +179,15 @@ export class MediaPage {
+	/* Modal mode */
+
+	/**
+	 *
```
Looks like this jsdoc header is missing as an FYI!

worldomonation replied:
Ah my bad, thanks!

dpasque commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -24,6 +26,10 @@ const selectors = {
 	imageEditorResetButton: 'button[data-e2e-button="reset"]',
 	imageEditorCancelButton: 'button[data-e2e-button="cancel"]',
+
+	// When embedded in a modal
+	modalActionButton: ( action: ModalAction ) =>
+		`button[data-e2e-button="${ action.toLowerCase() }"]`,
```
From my personal stylistic perspective, it feels a little bit off to have the modal handling in this media POM class. Maybe it violates the single responsibility principle? The modal dialog seems to me like something distinct from its contents. 

Two alternative suggestions:
- We might be able to just make a generic modal dialog component, as I believe we re-use the same modal DOM structure across Calypso.
- Alternatively, we could maybe make a class called MediaModal that has a property that gives access to the underlying media POM class contained within the modal?

What do you think?

worldomonation replied:
Looks like a comment I had left in this section of the code explaining my decision went missing. I'll summarize what I had written in it.

The gist was that I had decided to place the modal related actions within `MediaPage` for a couple of reasons:
- there are few, if any other places in Calypso that embed an entire Calypso page within a modal.
- I considered the modal to be just one another form in which the MediaPage can manifest itself.

My decision was driven primarily by the first point. While Calypso does use react modals throughout, we have so far created components for modals that are 'tailor made' for the component in question (eg. `PreviewComponent`, `RevisionComponent`).

With that said, if we are to change the implementation to a dialog modal component, then my vote would be on using dependency injection.

`const reactModalComponent = new ReactModalComponent( page, new MarketingPage(page) )`

then interact with the injected page object via:

`await reactModalComponent.pageObject.<method_name>`.

What do you think about these?

dpasque replied:
Yeah, that reasoning makes a lot of sense! I like the style of the dependency injection you l aid out -- what's cool too is we could use generic typing on the argument of the POM class passed in to have type safety on the methods on the `pageObject`.

I think we'll eventually need the ReactModalComponent, but we can definitely punt that until later and we need it more (and not do it in this PR). I'll defer to you! 🙂

worldomonation replied:
Addressed in 0b1be87bc91b305ae5ac6f91c7b4a865431d7c49.

dpasque reviewed (approved):
Looks great! 🚀  If you want to do any of the larger scope things here (if you agree with them), go for it, but otherwise I think we're good to go and we can follow-up on any of those in future PRs!

worldomonation committed:
Implement changes to interacting with react modals:
- create new `ReactModalComponent` class to interact with embedded modal pages.
- replace direct usage of MediaPage with `ReactMediaComponent` in ImageBlock.
- restructure spec to remove unnecessary base test case for image.
